### PR TITLE
[GCS]Add gcs resource manager

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
When GCS schedules actor or placement group, it needs to obtain the cluster resource view, so we add gcs resource manager.

## Related issue number
https://github.com/ray-project/ray/issues/12226

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
LGTM in general, but I wonder if we should start using the new classes for resources instead of `ResourceSet` and `ClusterResource`. cc @wuisawesome We are using the new resource classes for the new scheduler right?

**Comment by rkooo567** [comment]
@ffbin Please add a response to the question!

**Comment by wuisawesome** [comment]
It would be nice to move everything over to the new scheduler's structures especially if we're writing new components. I'm focusing more of my efforts on the raylet side though.

**Comment by rkooo567** [comment]
@wuisawesome Is new and old scheduler sharing the same heartbeat message in grpc? If so, this should be simple right?

**Comment by wuisawesome** [comment]
Yeah they're all serialized to the same protobuf format.

**Comment by ffbin** [comment]
@rkooo567 @wuisawesome  I will use the new classes for resources instead of `ResourceSet` and `ClusterResource`, thanks.

**Comment by ffbin** [comment]
Hi @rkooo567 , https://github.com/ray-project/ray/pull/12253 depends on this PR. How about start using the new classes for resources in next pr?

File: src/ray/gcs/gcs_server/gcs_resource_manager.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+/// It is used for actor and placement group scheduling.
+/// Non-thread safe.
+class GcsResourceManagerInterface {
+ public:
+  virtual ~GcsResourceManagerInterface() {}
+
+  /// Get the resources of all nodes in the cluster.
+  ///
+  /// \return The resources of all nodes in the cluster.
+  virtual const absl::flat_hash_map<NodeID, ResourceSet> &GetClusterResources() const = 0;
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_resource_manager.h:33]: Why not use `const absl::flat_hash_map<NodeID, SchedulingResources> &`

**Comment by wumuzi520** [comment]
I think it is necessary to move the internal GcsResourceManger directly to the community.

File: src/ray/gcs/gcs_server/gcs_resource_manager.h (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_resource_manager.h:33]: The resources of actor / task are not managed in GCS. We can only get the available resources from heartbeat.

**Comment by raulchen** [comment]
@wuisawesome @rkooo567 I think this PR is more like a refactor, instead of adding new feature. Fixing #12253  needs this PR because we need a module in GCS that manages resources. 

This shouldn't add new burden for the new scheduler migration.  because we just move `ResourceSet` from `GcsNodeManager` to `GcsResourceManager`.  

My suggestion is to combine this and #12253  in one PR, and change the data structure later. Is that okay?

**Comment by raulchen** [comment]
Btw, do we not have a common data structure interface? How are some tests runnable for both the old and the new scheduler then (given that GcsNodeManager is using ResourceSet)?

**Comment by ffbin** [comment]
Combine this pr and #12253 in one PR.

**Pull request closed without merging**